### PR TITLE
[Mobile] Long press transaction to reveal floating action bar with bulk actions 

### PR DESCRIPTION
**Pull request by joel-jeremy**
image
image
image

### CODE REVIEW

**Comment by youngcw** [comment]
If you bulk change the account on the transactions they stay selected even after they are visually gone. That's maybe ok but feels a bit weird.

**Comment by joel-jeremy** [comment]
@youngcw Updated to clear selection after the batch operations and added undo notifications since I realized these operetions can easily mess up the transactions if a mis click happened.

**Comment by youngcw** [comment]
Nice, I like the undo option.  Im thinking it would be nice to leave the transactions selected so other changes can be made.  Something like, edit the notes the change the category for a batch.  I do that on occasion.  Maybe we can leave this as is and see how it goes?  What do you think?

**Comment by joel-jeremy** [comment]
> Nice, I like the undo option. Im thinking it would be nice to leave the transactions selected so other changes can be made. Something like, edit the notes the change the category for a batch. I do that on occasion. Maybe we can leave this as is and see how it goes? What do you think?

Agreed, it makes sense to leave them selected. Updated the code :)

**Comment by joel-jeremy** [comment]
I also added a `useUndo` hook here which I plan to use in the budget page to show undo notification when applying goal templates.

**Comment by youngcw** [comment]
@MatissJanis @twk3 Could one of you look over this?  With the new hooks and such this deserves more experienced eyes.

File: packages/desktop-client/src/hooks/useUndo.ts
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useUndo.ts:33]: **🥜 nitpick:** ‏the async/await here is unnecessary

```suggestion
  const dispatchUndo = useCallback(() => {
    dispatch(undo());
  }, [dispatch]);

  const dispatchRedo = useCallback(() => {
    dispatch(redo());
  }, [dispatch]);
```

File: packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx:289]: **💭 thought:** ‏We definitely should do that! 😅

File: packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx:186]: **🥜 nitpick:** ‏by moving this constant definition outside the component you _might_ reduce the amonut of renders due to the object being re-created each time.

File: packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx:244]: **💭 thought:** ‏What happens if `onClearSelectedTransactions` is undefined? The button is still visible, but nothing happens when clicked? That would be quite bad. Perhaps we can wrap the button in a conditional OR make the `onClearSelectedTransactions` prop mandatory?

File: packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx:264]: **👏 praise:** ‏in this case the aria-label makes total sense :)

**Review by MatissJanis — COMMENTED** [review]
Overall LGTM! 

Small bug (not sure if related to this PR though): "link schedule" opens a modal, but after you select a schedule to link - the modal stays open. So it _feels_ like nothing happened.

Really like the thought that went into the undo feature! That's super nice. Though I find the undo notification quite disruptive. What if we instead added a "undo" button besides the "edit" button (maybe with a undo icon instead? just a thought)?

Would also be really nice to support mobile _gestures_ for undo action (i.e. you can shake iPhone to undo stuff). But that's more of a feature request. :)

File: packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx
Comment (joel-jeremy) [reply packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx]: Updated. :) I'm thinking about whether we want to enable batch editing in the category transactions page. Thoughts?

Comment (MatissJanis) [reply packages/desktop-client/src/components/mobile/transactions/TransactionList.jsx]: Ideally we would have equal feature-set in all the pages where we list transactions. So from that perspective - I think it makes sense to also enable in category transaction page.

**Comment by MatissJanis** [comment]
Code looks good :+1:

Not sure if you missed [this comment](https://github.com/actualbudget/actual/pull/2892#pullrequestreview-2183830052) with a bug report + a feedback item on the "undo" notifications.

**Comment by joel-jeremy** [comment]
> Small bug (not sure if related to this PR though): "link schedule" opens a modal, but after you select a schedule to link - the modal stays open. So it feels like nothing happened.

I can't seem to replicate this one. Maybe the recent commits fixed it?

> Really like the thought that went into the undo feature! That's super nice. Though I find the undo notification quite disruptive. What if we instead added a "undo" button besides the "edit" button (maybe with a undo icon instead? just a thought)?
I'm planning to reuse the undo notification on the budget page too so I would like to keep the "undo" mechanism consistent across the pages for now. Once we have that running, we can definitely revamp the undo notification to be less disruptive or maybe we can just use less text on the notification so it takes up less screen real estate?

**Comment by MatissJanis** [comment]
This is what I see:

https://github.com/user-attachments/assets/b9ddb896-ed9e-4a3f-949e-8816379f0232

And now that I think about it more - there's also no "undo" notification for this action. But there is one for un-linking which is strange.

> I'm planning to reuse the undo notification on the budget page too (https://github.com/actualbudget/actual/pull/3085) so I would like to keep the "undo" mechanism consistent across the pages for now. Once we have that running, we can definitely revamp the undo notification to be less disruptive or maybe we can just use less text on the notification so it takes up less screen real estate?

Less text works fine. And then we can nicely position it to be above the purple action bar. (+ would be really nice to auto-close it after a delay).

**Comment by Teprifer** [comment]
Yay more functionality! I did pick up two niggles when trying it out.

1) The edit notes Prepend and Append options give an incorrect undo notification as the wording indicates the note from one of the selected is what all were set to. When in reality each note add the text prepended/appended to their existing content.

Example:
Pre-edit notes for the 3 transactions were in turn: 
* Test one
* Test Two
* Test Three

Edit notes Prepend 'prepend'
Result:
image


---

2) The Edit Amount dialog brings up the qwerty keyboard and not the numbers keyboard(Android, firefox incognito mode). Obviously this generates a red something internally went wrong notice if text is attempted to be entered.

**Comment by joel-jeremy** [comment]
Thank you for your feedback @Teprifer. I have fixed #1. I guess for number we're going to have to address that in a separate PR. Maybe I should disable bulk edit for amount for now - just like Date is currently disabled. WDYT?

File: packages/loot-core/src/types/models/transaction.d.ts
Comment (joel-jeremy) [inline packages/loot-core/src/types/models/transaction.d.ts:29]: I had to fix these type to properly strict type the new `useTransactionBatchActions` hook

**Comment by Teprifer** [comment]
> Thank you for your feedback @Teprifer. I have fixed #1. I guess for number we're going to have to address that in a separate PR. Maybe I should disable bulk edit for amount for now - just like Date is currently disabled. WDYT?

Cheers, not strongly opinionated, although I edge slightly towards your suggestion of disabling as errors are unfriendly.

**Comment by MatissJanis** [comment]
Sorry, but I'm still **really** disliking the undo notification UI. As a minimum - could we make it so it doesn't overlay the action bar? To push it up.

Also I think the close delay should be increased to something much higher (~10s?). To give people time to react before it closes.

**Comment by joel-jeremy** [comment]
@MatissJanis Updated :)

**Comment by MatissJanis** [comment]
Thanks! The positioning is much better now. But now we have one more bug: if you perform an action - the undo notification opens up. But there is no way to close it. And without closing it - the entire page is not usable (clicks are blocked everywhere). So you have to wait for the notification to dissapear.

Would it be possible to allow clicking in the action bar if the notification is open? And perhaps to also add a "x" for the undo notification so it can be dismissed more quickly if the user wants to.

**Comment by joel-jeremy** [comment]
Floating action bar should now be clickable even if a notification is present :)

**Comment by Teprifer** [comment]
Appears to be temporarily not working? Can't long press or tap to edit/view a transaction.

Browser console:
image

**Comment by joel-jeremy** [comment]
@Teprifer Nice catch! I went into a deep dive into react aria's codebase and found out that one needs to use `PressResponder` to handle both press and long press on react-aria-components's `Button` component

**Review by MatissJanis — DISMISSED** [review]
LGTM on my end.

Please wait for @Teprifer approval too before merging since it's quite a big change :)

**Review by Teprifer — DISMISSED** [review]
The afore mentioned issue is fixed and a quick run through of the options checks out. :)

I have identified a minor visual issue: If you long press to select, then single tap to unselect or X out the selection using the bottom bar, the unselected transactions have a white/gold indicator on the left side where the purple selected indicator was.
This goes away after navigation off page and back.

Most visible in mobile (images from Firefox, but confirmed in android chrome too):
image
image

From desktop chrome:
image

I think the room for confusion is relatively minor so I'll approve with the caveat @MatissJanis confirms if creating a follow-up is ok.


_I initially had an issue where transactions were still selected even after navigating off the account page and to another account, then back again. But I wasn't able to replicate this._

**Review by Teprifer — DISMISSED** [review]
If you work any faster I might start thinking you were sitting at your keyboard just waiting. ;)

Confirmed issue fixed (visual check only).

File: packages/desktop-client/e2e/rules.test.js-snapshots/Rules-creates-a-rule-and-makes-sure-it-is-applied-when-creating-a-transaction-1-chromium-linux.png
Comment (MatissJanis) [inline packages/desktop-client/e2e/rules.test.js-snapshots/Rules-creates-a-rule-and-makes-sure-it-is-applied-when-creating-a-transaction-1-chromium-linux.png:1]: Why did this change?

Comment (joel-jeremy) [reply packages/desktop-client/e2e/rules.test.js-snapshots/Rules-creates-a-rule-and-makes-sure-it-is-applied-when-creating-a-transaction-1-chromium-linux.png:1]: Hmm not sure what's causing this change. There are no updates to that component on this PR.

Comment (MatissJanis) [reply packages/desktop-client/e2e/rules.test.js-snapshots/Rules-creates-a-rule-and-makes-sure-it-is-applied-when-creating-a-transaction-1-chromium-linux.png:1]: Patched the VRT issue here: https://github.com/actualbudget/actual/pull/3186

Would you mind updating your branch? There should be no changes in VRT here after you update the branch :)

Comment (joel-jeremy) [reply packages/desktop-client/e2e/rules.test.js-snapshots/Rules-creates-a-rule-and-makes-sure-it-is-applied-when-creating-a-transaction-1-chromium-linux.png:1]: Thanks for the patch! I updated the branch and it's now good :)

**Review by MatissJanis — APPROVED** [review]

**Pull request merged**
